### PR TITLE
Make submission of the course verification deadline optional

### PR DESCRIPTION
**Pull request by rlucioni**
Verification deadline should only be required if the course actually requires verification. XCOM-537.

@clintonb @jimabramson

### CODE REVIEW

**Comment by clintonb** [comment]
The field should be required if a course contains a verified course seat.

File: ecommerce/extensions/api/v2/tests/views/test_publication.py [hunk 1]
```diff
@@ -238,3 +238,14 @@ def test_missing_product_price(self):
         response = self.client.post(self.create_path, json.dumps(self.data), JSON_CONTENT_TYPE)
         self.assertEqual(response.status_code, 400)
         self._assert_course_saved(self.course_id)
+
+    def test_verification_deadline_optionalz(self):
```
Comment (jimabramson) [inline ecommerce/extensions/api/v2/tests/views/test_publication.py]: s/optionalz/optional  (or, update all other tests to also end with a z)

**Comment by jimabramson** [comment]
@clintonb to clarify - is it always required if id verification is required, or only required with verified cert?  also, should it be allowed to be set if verification is not required, since irrelevant?

**Comment by rlucioni** [comment]
@jimabramson I'm implementing it as if it's the former.

File: ecommerce/extensions/api/v2/tests/views/test_publication.py (hunk 1 above)
Comment (rlucioni) [inline ecommerce/extensions/api/v2/tests/views/test_publication.py]: Updated; resisted the temptation to end all other tests with a z.

**Comment by rlucioni** [comment]
@clintonb updated.

**Comment by clintonb** [comment]
@jimabramson @rlucioni the updated validation follows my understanding which is that if any seats require ID verification, a verification deadline must be set. Non-verified courses do not require it.

**Comment by clintonb** [comment]
:+1:

**Comment by wedaly** [comment]
@rlucioni Really sorry to bring this up so late in the game.  While it's true that currently all courses with a verified seat have a verification deadline, the primary motivation for separating the dates was so that we could disable the deadline for credit courses.  The thought is that in-course reverification can take the place of a hard verification deadline.

I'm not sure if this is a blocker or not, since we're configuring the first run of credit courses manually.

**Comment by wedaly** [comment]
In sum, I think the validation rules should be:
- Only courses with a verified seat are allowed to have a verification deadline.
- For courses with a verified seat, the verification deadline is optional.
- If a course has both an upgrade and verification deadline set, then the verification deadline must be after the upgrade deadline.

where "verified seat" currently means either a "verified" or "professional" (with verification!) course mode.

File: ecommerce/extensions/api/serializers.py [hunk 2]
```diff
@@ -158,7 +158,8 @@ class AtomicPublicationSerializer(serializers.Serializer):  # pylint: disable=ab
     """
     id = serializers.RegexField(COURSE_ID_REGEX, max_length=255)
     name = serializers.CharField(max_length=255)
-    verification_deadline = serializers.DateTimeField()
+    # Verification deadline should only be required if the course actually requires verification.
+    verification_deadline = serializers.DateTimeField(required=False)
```
Comment (clintonb) [inline ecommerce/extensions/api/serializers.py]: Also include `allow_null=True`.

Comment (rlucioni) [inline ecommerce/extensions/api/serializers.py]: Updated.

**Comment by rlucioni** [comment]
@wedaly I'm not comfortable rushing to implement validation rules that I don't fully understand. Given that 1) it's late in the game, 2) the validation you're describing is already in place on edx-platform, and 3) adding these validation rules to the CAT isn't critical to the migration, would it be okay if we address this kind of validation in a separate story?

Since this validation is present on edx-platform (specifically, within the Commerce API), a user won't be able to, for example, set a verification deadline earlier than a seat's upgrade deadline. If this were attempted, publication to the LMS would fail and the data wouldn't be persisted.

**Comment by wedaly** [comment]
@rlucioni That's fine as long as we have a way to disable the verification deadline (and ensure it stays disabled) for credit courses.

**Comment by clintonb** [comment]
:+1:

**Comment by jimabramson** [comment]
:+1:

**Comment by rlucioni** [comment]
FYI, I've created [XCOM-544](https://openedx.atlassian.net/browse/XCOM-544) to track the addition of the validation rules described above.

**Pull request merged**